{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create an interactive Parallel Plot\n",
    "To demonstrate the use of the interactive parallel plot, we use a project already loaded into the CKG database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n",
      "/Users/plh450/CKG_env_updated/lib/python3.6/site-packages/rpy2/robjects/pandas2ri.py:197: UserWarning:\n",
      "\n",
      "R object inheriting from \"POSIXct\" but without attribute \"tzone\".\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "        <script type=\"text/javascript\">\n",
       "        window.PlotlyConfig = {MathJaxConfig: 'local'};\n",
       "        if (window.MathJax) {MathJax.Hub.Config({SVG: {font: \"STIX-Web\"}});}\n",
       "        if (typeof require !== 'undefined') {\n",
       "        require.undef(\"plotly\");\n",
       "        requirejs.config({\n",
       "            paths: {\n",
       "                'plotly': ['https://cdn.plot.ly/plotly-latest.min']\n",
       "            }\n",
       "        });\n",
       "        require(['plotly'], function(Plotly) {\n",
       "            window._Plotly = Plotly;\n",
       "        });\n",
       "        }\n",
       "        </script>\n",
       "        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from report_manager import project, dataset, report\n",
    "from analytics_core.viz import viz as plots\n",
    "import networkx as nx\n",
    "from networkx.readwrite import json_graph\n",
    "from plotly.offline import init_notebook_mode, iplot\n",
    "import plotly.graph_objs as go\n",
    "from scipy.stats import zscore\n",
    "init_notebook_mode(connected=True)\n",
    "%matplotlib inline\n",
    "import ipywidgets as widgets\n",
    "from ipywidgets import interact, interact_manual\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "import logging\n",
    "logger = logging.getLogger()\n",
    "logger.setLevel(logging.CRITICAL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### We create a new project object and load the respective data and report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "my_project = project.Project(identifier='P0000001', datasets={}, report={})\n",
    "my_project.load_project_data()\n",
    "my_project.load_project_report()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### We can now access to all the results for each data type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['multiomics', 'clinical', 'proteomics'])"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_project.list_datasets()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### We will use the results from the proteomics analyses. We access the dataset 'proteomics' for further analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "proteomics_dataset = my_project.get_dataset('proteomics')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### The available analysis for this dataset are:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['complex_associations',\n",
       " 'correlation_correlation',\n",
       " 'disease_associations',\n",
       " 'drug_associations',\n",
       " 'go annotation',\n",
       " 'go_enrichment_Biological_processes_regulation_enrichment',\n",
       " 'interaction_network',\n",
       " 'literature_associations_publications_abstracts',\n",
       " 'number of modified proteins',\n",
       " 'number of peptides',\n",
       " 'number of proteins',\n",
       " 'original',\n",
       " 'overview statistics_summary',\n",
       " 'pathway annotation',\n",
       " 'pathway_enrichment_Pathways_regulation_enrichment',\n",
       " 'processed',\n",
       " 'protein biomarkers',\n",
       " 'regulated',\n",
       " 'regulation table']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_project.get_dataset('proteomics').list_dataframes()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### We can access the different dataframes like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annotation</th>\n",
       "      <th>group</th>\n",
       "      <th>identifier</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>endosomal transport</td>\n",
       "      <td>None</td>\n",
       "      <td>DPY30~Q9C005</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>regulation of megakaryocyte differentiation</td>\n",
       "      <td>None</td>\n",
       "      <td>DPY30~Q9C005</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>histone H3-K4 methylation</td>\n",
       "      <td>None</td>\n",
       "      <td>DPY30~Q9C005</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>proteolysis</td>\n",
       "      <td>background</td>\n",
       "      <td>CPQ~Q9Y646</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>thyroid hormone generation</td>\n",
       "      <td>background</td>\n",
       "      <td>CPQ~Q9Y646</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>tissue regeneration</td>\n",
       "      <td>background</td>\n",
       "      <td>CPQ~Q9Y646</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>peptide catabolic process</td>\n",
       "      <td>background</td>\n",
       "      <td>CPQ~Q9Y646</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>complement activation, alternative pathway</td>\n",
       "      <td>background</td>\n",
       "      <td>CFHR5~Q9BXR6</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>regulation of complement activation</td>\n",
       "      <td>background</td>\n",
       "      <td>CFHR5~Q9BXR6</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>negative regulation of protein binding</td>\n",
       "      <td>background</td>\n",
       "      <td>CFHR5~Q9BXR6</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>positive regulation of cytolysis</td>\n",
       "      <td>background</td>\n",
       "      <td>CFHR5~Q9BXR6</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>skeletal system development</td>\n",
       "      <td>None</td>\n",
       "      <td>BCAN~Q96GW7</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>cell adhesion</td>\n",
       "      <td>None</td>\n",
       "      <td>BCAN~Q96GW7</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>central nervous system development</td>\n",
       "      <td>None</td>\n",
       "      <td>BCAN~Q96GW7</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>hippocampus development</td>\n",
       "      <td>None</td>\n",
       "      <td>BCAN~Q96GW7</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>extracellular matrix organization</td>\n",
       "      <td>None</td>\n",
       "      <td>BCAN~Q96GW7</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chondroitin sulfate biosynthetic process</td>\n",
       "      <td>None</td>\n",
       "      <td>BCAN~Q96GW7</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chondroitin sulfate catabolic process</td>\n",
       "      <td>None</td>\n",
       "      <td>BCAN~Q96GW7</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>dermatan sulfate biosynthetic process</td>\n",
       "      <td>None</td>\n",
       "      <td>BCAN~Q96GW7</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>synapse maturation</td>\n",
       "      <td>None</td>\n",
       "      <td>BCAN~Q96GW7</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>cell adhesion</td>\n",
       "      <td>background</td>\n",
       "      <td>CNTN3~Q9P232</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>nervous system development</td>\n",
       "      <td>background</td>\n",
       "      <td>CNTN3~Q9P232</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>cellular protein modification process</td>\n",
       "      <td>background</td>\n",
       "      <td>ST3GAL6~Q9Y274</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>protein glycosylation</td>\n",
       "      <td>background</td>\n",
       "      <td>ST3GAL6~Q9Y274</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>glycolipid metabolic process</td>\n",
       "      <td>background</td>\n",
       "      <td>ST3GAL6~Q9Y274</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>glycolipid biosynthetic process</td>\n",
       "      <td>background</td>\n",
       "      <td>ST3GAL6~Q9Y274</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>oligosaccharide metabolic process</td>\n",
       "      <td>background</td>\n",
       "      <td>ST3GAL6~Q9Y274</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>keratan sulfate biosynthetic process</td>\n",
       "      <td>background</td>\n",
       "      <td>ST3GAL6~Q9Y274</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>cellular response to interleukin-6</td>\n",
       "      <td>background</td>\n",
       "      <td>ST3GAL6~Q9Y274</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>sialylation</td>\n",
       "      <td>background</td>\n",
       "      <td>ST3GAL6~Q9Y274</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17268</th>\n",
       "      <td>cerebellar Purkinje cell layer development</td>\n",
       "      <td>None</td>\n",
       "      <td>AARS~P49588</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17269</th>\n",
       "      <td>negative regulation of neuron apoptotic process</td>\n",
       "      <td>None</td>\n",
       "      <td>AARS~P49588</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17270</th>\n",
       "      <td>neuromuscular process controlling balance</td>\n",
       "      <td>None</td>\n",
       "      <td>AARS~P49588</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17271</th>\n",
       "      <td>aminoacyl-tRNA metabolism involved in translat...</td>\n",
       "      <td>None</td>\n",
       "      <td>AARS~P49588</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17272</th>\n",
       "      <td>regulation of cytoplasmic translational fidelity</td>\n",
       "      <td>None</td>\n",
       "      <td>AARS~P49588</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17273</th>\n",
       "      <td>chromatin silencing at rDNA</td>\n",
       "      <td>None</td>\n",
       "      <td>HIST1H3A~P68431</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17274</th>\n",
       "      <td>chromatin organization</td>\n",
       "      <td>None</td>\n",
       "      <td>HIST1H3A~P68431</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17275</th>\n",
       "      <td>nucleosome assembly</td>\n",
       "      <td>None</td>\n",
       "      <td>HIST1H3A~P68431</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17276</th>\n",
       "      <td>DNA replication-dependent nucleosome assembly</td>\n",
       "      <td>None</td>\n",
       "      <td>HIST1H3A~P68431</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17277</th>\n",
       "      <td>blood coagulation</td>\n",
       "      <td>None</td>\n",
       "      <td>HIST1H3A~P68431</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17278</th>\n",
       "      <td>telomere organization</td>\n",
       "      <td>None</td>\n",
       "      <td>HIST1H3A~P68431</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17279</th>\n",
       "      <td>interleukin-7-mediated signaling pathway</td>\n",
       "      <td>None</td>\n",
       "      <td>HIST1H3A~P68431</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17280</th>\n",
       "      <td>cellular protein metabolic process</td>\n",
       "      <td>None</td>\n",
       "      <td>HIST1H3A~P68431</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17281</th>\n",
       "      <td>regulation of megakaryocyte differentiation</td>\n",
       "      <td>None</td>\n",
       "      <td>HIST1H3A~P68431</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17282</th>\n",
       "      <td>negative regulation of gene expression, epigen...</td>\n",
       "      <td>None</td>\n",
       "      <td>HIST1H3A~P68431</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17283</th>\n",
       "      <td>protein heterotetramerization</td>\n",
       "      <td>None</td>\n",
       "      <td>HIST1H3A~P68431</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17284</th>\n",
       "      <td>regulation of gene silencing by miRNA</td>\n",
       "      <td>None</td>\n",
       "      <td>HIST1H3A~P68431</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17285</th>\n",
       "      <td>regulation of gene silencing</td>\n",
       "      <td>None</td>\n",
       "      <td>HIST1H3A~P68431</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17286</th>\n",
       "      <td>cytoskeleton organization</td>\n",
       "      <td>None</td>\n",
       "      <td>KRT4~P19013</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17287</th>\n",
       "      <td>epithelial cell differentiation</td>\n",
       "      <td>None</td>\n",
       "      <td>KRT4~P19013</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17288</th>\n",
       "      <td>keratinization</td>\n",
       "      <td>None</td>\n",
       "      <td>KRT4~P19013</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17289</th>\n",
       "      <td>negative regulation of epithelial cell prolife...</td>\n",
       "      <td>None</td>\n",
       "      <td>KRT4~P19013</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17290</th>\n",
       "      <td>cornification</td>\n",
       "      <td>None</td>\n",
       "      <td>KRT4~P19013</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17291</th>\n",
       "      <td>signal transduction</td>\n",
       "      <td>None</td>\n",
       "      <td>SIRPB1~O00241</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17292</th>\n",
       "      <td>cell surface receptor signaling pathway</td>\n",
       "      <td>None</td>\n",
       "      <td>SIRPB1~O00241</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17293</th>\n",
       "      <td>positive regulation of cell-cell adhesion</td>\n",
       "      <td>None</td>\n",
       "      <td>SIRPB1~O00241</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17294</th>\n",
       "      <td>neutrophil degranulation</td>\n",
       "      <td>None</td>\n",
       "      <td>SIRPB1~O00241</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17295</th>\n",
       "      <td>innate immune response</td>\n",
       "      <td>None</td>\n",
       "      <td>SIRPB1~O00241</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17296</th>\n",
       "      <td>positive regulation of phagocytosis</td>\n",
       "      <td>None</td>\n",
       "      <td>SIRPB1~O00241</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17297</th>\n",
       "      <td>positive regulation of T cell activation</td>\n",
       "      <td>None</td>\n",
       "      <td>SIRPB1~O00241</td>\n",
       "      <td>UniProt</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17298 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              annotation       group  \\\n",
       "0                                    endosomal transport        None   \n",
       "1            regulation of megakaryocyte differentiation        None   \n",
       "2                              histone H3-K4 methylation        None   \n",
       "3                                            proteolysis  background   \n",
       "4                             thyroid hormone generation  background   \n",
       "5                                    tissue regeneration  background   \n",
       "6                              peptide catabolic process  background   \n",
       "7             complement activation, alternative pathway  background   \n",
       "8                    regulation of complement activation  background   \n",
       "9                 negative regulation of protein binding  background   \n",
       "10                      positive regulation of cytolysis  background   \n",
       "11                           skeletal system development        None   \n",
       "12                                         cell adhesion        None   \n",
       "13                    central nervous system development        None   \n",
       "14                               hippocampus development        None   \n",
       "15                     extracellular matrix organization        None   \n",
       "16              chondroitin sulfate biosynthetic process        None   \n",
       "17                 chondroitin sulfate catabolic process        None   \n",
       "18                 dermatan sulfate biosynthetic process        None   \n",
       "19                                    synapse maturation        None   \n",
       "20                                         cell adhesion  background   \n",
       "21                            nervous system development  background   \n",
       "22                 cellular protein modification process  background   \n",
       "23                                 protein glycosylation  background   \n",
       "24                          glycolipid metabolic process  background   \n",
       "25                       glycolipid biosynthetic process  background   \n",
       "26                     oligosaccharide metabolic process  background   \n",
       "27                  keratan sulfate biosynthetic process  background   \n",
       "28                    cellular response to interleukin-6  background   \n",
       "29                                           sialylation  background   \n",
       "...                                                  ...         ...   \n",
       "17268         cerebellar Purkinje cell layer development        None   \n",
       "17269    negative regulation of neuron apoptotic process        None   \n",
       "17270          neuromuscular process controlling balance        None   \n",
       "17271  aminoacyl-tRNA metabolism involved in translat...        None   \n",
       "17272   regulation of cytoplasmic translational fidelity        None   \n",
       "17273                        chromatin silencing at rDNA        None   \n",
       "17274                             chromatin organization        None   \n",
       "17275                                nucleosome assembly        None   \n",
       "17276      DNA replication-dependent nucleosome assembly        None   \n",
       "17277                                  blood coagulation        None   \n",
       "17278                              telomere organization        None   \n",
       "17279           interleukin-7-mediated signaling pathway        None   \n",
       "17280                 cellular protein metabolic process        None   \n",
       "17281        regulation of megakaryocyte differentiation        None   \n",
       "17282  negative regulation of gene expression, epigen...        None   \n",
       "17283                      protein heterotetramerization        None   \n",
       "17284              regulation of gene silencing by miRNA        None   \n",
       "17285                       regulation of gene silencing        None   \n",
       "17286                          cytoskeleton organization        None   \n",
       "17287                    epithelial cell differentiation        None   \n",
       "17288                                     keratinization        None   \n",
       "17289  negative regulation of epithelial cell prolife...        None   \n",
       "17290                                      cornification        None   \n",
       "17291                                signal transduction        None   \n",
       "17292            cell surface receptor signaling pathway        None   \n",
       "17293          positive regulation of cell-cell adhesion        None   \n",
       "17294                           neutrophil degranulation        None   \n",
       "17295                             innate immune response        None   \n",
       "17296                positive regulation of phagocytosis        None   \n",
       "17297           positive regulation of T cell activation        None   \n",
       "\n",
       "            identifier   source  \n",
       "0         DPY30~Q9C005  UniProt  \n",
       "1         DPY30~Q9C005  UniProt  \n",
       "2         DPY30~Q9C005  UniProt  \n",
       "3           CPQ~Q9Y646  UniProt  \n",
       "4           CPQ~Q9Y646  UniProt  \n",
       "5           CPQ~Q9Y646  UniProt  \n",
       "6           CPQ~Q9Y646  UniProt  \n",
       "7         CFHR5~Q9BXR6  UniProt  \n",
       "8         CFHR5~Q9BXR6  UniProt  \n",
       "9         CFHR5~Q9BXR6  UniProt  \n",
       "10        CFHR5~Q9BXR6  UniProt  \n",
       "11         BCAN~Q96GW7  UniProt  \n",
       "12         BCAN~Q96GW7  UniProt  \n",
       "13         BCAN~Q96GW7  UniProt  \n",
       "14         BCAN~Q96GW7  UniProt  \n",
       "15         BCAN~Q96GW7  UniProt  \n",
       "16         BCAN~Q96GW7  UniProt  \n",
       "17         BCAN~Q96GW7  UniProt  \n",
       "18         BCAN~Q96GW7  UniProt  \n",
       "19         BCAN~Q96GW7  UniProt  \n",
       "20        CNTN3~Q9P232  UniProt  \n",
       "21        CNTN3~Q9P232  UniProt  \n",
       "22      ST3GAL6~Q9Y274  UniProt  \n",
       "23      ST3GAL6~Q9Y274  UniProt  \n",
       "24      ST3GAL6~Q9Y274  UniProt  \n",
       "25      ST3GAL6~Q9Y274  UniProt  \n",
       "26      ST3GAL6~Q9Y274  UniProt  \n",
       "27      ST3GAL6~Q9Y274  UniProt  \n",
       "28      ST3GAL6~Q9Y274  UniProt  \n",
       "29      ST3GAL6~Q9Y274  UniProt  \n",
       "...                ...      ...  \n",
       "17268      AARS~P49588  UniProt  \n",
       "17269      AARS~P49588  UniProt  \n",
       "17270      AARS~P49588  UniProt  \n",
       "17271      AARS~P49588  UniProt  \n",
       "17272      AARS~P49588  UniProt  \n",
       "17273  HIST1H3A~P68431  UniProt  \n",
       "17274  HIST1H3A~P68431  UniProt  \n",
       "17275  HIST1H3A~P68431  UniProt  \n",
       "17276  HIST1H3A~P68431  UniProt  \n",
       "17277  HIST1H3A~P68431  UniProt  \n",
       "17278  HIST1H3A~P68431  UniProt  \n",
       "17279  HIST1H3A~P68431  UniProt  \n",
       "17280  HIST1H3A~P68431  UniProt  \n",
       "17281  HIST1H3A~P68431  UniProt  \n",
       "17282  HIST1H3A~P68431  UniProt  \n",
       "17283  HIST1H3A~P68431  UniProt  \n",
       "17284  HIST1H3A~P68431  UniProt  \n",
       "17285  HIST1H3A~P68431  UniProt  \n",
       "17286      KRT4~P19013  UniProt  \n",
       "17287      KRT4~P19013  UniProt  \n",
       "17288      KRT4~P19013  UniProt  \n",
       "17289      KRT4~P19013  UniProt  \n",
       "17290      KRT4~P19013  UniProt  \n",
       "17291    SIRPB1~O00241  UniProt  \n",
       "17292    SIRPB1~O00241  UniProt  \n",
       "17293    SIRPB1~O00241  UniProt  \n",
       "17294    SIRPB1~O00241  UniProt  \n",
       "17295    SIRPB1~O00241  UniProt  \n",
       "17296    SIRPB1~O00241  UniProt  \n",
       "17297    SIRPB1~O00241  UniProt  \n",
       "\n",
       "[17298 rows x 4 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_project.get_dataset('proteomics').get_dataframe('go annotation')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### In this case, we will use the the processed dataframe with transformed and imputed LFQ intensities. We then normalize the data using Z Score."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "proteomics_dataset = my_project.get_dataset('proteomics')\n",
    "processed_df = proteomics_dataset.get_dataframe('processed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A2M~P01023</th>\n",
       "      <th>A30~A2MYE2</th>\n",
       "      <th>ABI3BP~Q7Z7G0</th>\n",
       "      <th>ACE~P12821</th>\n",
       "      <th>ACTB~P60709</th>\n",
       "      <th>ACTN1~P12814</th>\n",
       "      <th>ADA2~Q9NZK5</th>\n",
       "      <th>ADAMTS13~Q76LX8</th>\n",
       "      <th>ADAMTSL4~Q6UY14</th>\n",
       "      <th>ADH4~P08319</th>\n",
       "      <th>...</th>\n",
       "      <th>VIM~P08670</th>\n",
       "      <th>VK3~A2N2F4</th>\n",
       "      <th>VNN1~O95497</th>\n",
       "      <th>VTN~P04004</th>\n",
       "      <th>VWF~P04275</th>\n",
       "      <th>YWHAZ~P63104</th>\n",
       "      <th>group</th>\n",
       "      <th>sample</th>\n",
       "      <th>scFv~Q65ZC9</th>\n",
       "      <th>subject</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>38.005877</td>\n",
       "      <td>28.173504</td>\n",
       "      <td>21.548283</td>\n",
       "      <td>22.173995</td>\n",
       "      <td>27.090330</td>\n",
       "      <td>25.039968</td>\n",
       "      <td>23.442151</td>\n",
       "      <td>24.010605</td>\n",
       "      <td>25.085820</td>\n",
       "      <td>23.389032</td>\n",
       "      <td>...</td>\n",
       "      <td>24.178889</td>\n",
       "      <td>25.835908</td>\n",
       "      <td>22.480055</td>\n",
       "      <td>32.815815</td>\n",
       "      <td>28.922779</td>\n",
       "      <td>19.205044</td>\n",
       "      <td>Cirrhosis</td>\n",
       "      <td>AS65</td>\n",
       "      <td>27.788928</td>\n",
       "      <td>S65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>37.306744</td>\n",
       "      <td>27.981907</td>\n",
       "      <td>27.342062</td>\n",
       "      <td>23.847270</td>\n",
       "      <td>27.461155</td>\n",
       "      <td>25.896268</td>\n",
       "      <td>23.754503</td>\n",
       "      <td>24.135818</td>\n",
       "      <td>19.190441</td>\n",
       "      <td>22.148706</td>\n",
       "      <td>...</td>\n",
       "      <td>23.709777</td>\n",
       "      <td>25.004889</td>\n",
       "      <td>23.852908</td>\n",
       "      <td>32.722121</td>\n",
       "      <td>29.881279</td>\n",
       "      <td>22.141285</td>\n",
       "      <td>Cirrhosis</td>\n",
       "      <td>AS66</td>\n",
       "      <td>26.869972</td>\n",
       "      <td>S66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>37.381169</td>\n",
       "      <td>28.857627</td>\n",
       "      <td>20.126953</td>\n",
       "      <td>22.863630</td>\n",
       "      <td>27.929764</td>\n",
       "      <td>24.295225</td>\n",
       "      <td>23.359443</td>\n",
       "      <td>24.121788</td>\n",
       "      <td>24.923476</td>\n",
       "      <td>23.017163</td>\n",
       "      <td>...</td>\n",
       "      <td>23.599064</td>\n",
       "      <td>26.271650</td>\n",
       "      <td>24.232132</td>\n",
       "      <td>32.755752</td>\n",
       "      <td>29.444625</td>\n",
       "      <td>18.872159</td>\n",
       "      <td>Cirrhosis</td>\n",
       "      <td>AS67</td>\n",
       "      <td>28.069328</td>\n",
       "      <td>S67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>38.416989</td>\n",
       "      <td>28.978380</td>\n",
       "      <td>25.501910</td>\n",
       "      <td>22.992774</td>\n",
       "      <td>27.152479</td>\n",
       "      <td>25.231288</td>\n",
       "      <td>23.701340</td>\n",
       "      <td>24.568309</td>\n",
       "      <td>24.878802</td>\n",
       "      <td>26.388112</td>\n",
       "      <td>...</td>\n",
       "      <td>24.179076</td>\n",
       "      <td>25.929200</td>\n",
       "      <td>24.269047</td>\n",
       "      <td>32.714014</td>\n",
       "      <td>29.397176</td>\n",
       "      <td>22.216971</td>\n",
       "      <td>Cirrhosis</td>\n",
       "      <td>AS69</td>\n",
       "      <td>28.170209</td>\n",
       "      <td>S69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>37.474782</td>\n",
       "      <td>28.748744</td>\n",
       "      <td>21.912284</td>\n",
       "      <td>22.209683</td>\n",
       "      <td>27.537048</td>\n",
       "      <td>22.392992</td>\n",
       "      <td>22.406264</td>\n",
       "      <td>24.961173</td>\n",
       "      <td>22.001683</td>\n",
       "      <td>24.339540</td>\n",
       "      <td>...</td>\n",
       "      <td>23.865224</td>\n",
       "      <td>26.701340</td>\n",
       "      <td>20.760923</td>\n",
       "      <td>32.722691</td>\n",
       "      <td>28.540895</td>\n",
       "      <td>20.886069</td>\n",
       "      <td>Cirrhosis</td>\n",
       "      <td>AS70</td>\n",
       "      <td>28.612280</td>\n",
       "      <td>S70</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 522 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   A2M~P01023  A30~A2MYE2  ABI3BP~Q7Z7G0  ACE~P12821  ACTB~P60709  \\\n",
       "0   38.005877   28.173504      21.548283   22.173995    27.090330   \n",
       "1   37.306744   27.981907      27.342062   23.847270    27.461155   \n",
       "2   37.381169   28.857627      20.126953   22.863630    27.929764   \n",
       "3   38.416989   28.978380      25.501910   22.992774    27.152479   \n",
       "4   37.474782   28.748744      21.912284   22.209683    27.537048   \n",
       "\n",
       "   ACTN1~P12814  ADA2~Q9NZK5  ADAMTS13~Q76LX8  ADAMTSL4~Q6UY14  ADH4~P08319  \\\n",
       "0     25.039968    23.442151        24.010605        25.085820    23.389032   \n",
       "1     25.896268    23.754503        24.135818        19.190441    22.148706   \n",
       "2     24.295225    23.359443        24.121788        24.923476    23.017163   \n",
       "3     25.231288    23.701340        24.568309        24.878802    26.388112   \n",
       "4     22.392992    22.406264        24.961173        22.001683    24.339540   \n",
       "\n",
       "   ...  VIM~P08670  VK3~A2N2F4  VNN1~O95497  VTN~P04004  VWF~P04275  \\\n",
       "0  ...   24.178889   25.835908    22.480055   32.815815   28.922779   \n",
       "1  ...   23.709777   25.004889    23.852908   32.722121   29.881279   \n",
       "2  ...   23.599064   26.271650    24.232132   32.755752   29.444625   \n",
       "3  ...   24.179076   25.929200    24.269047   32.714014   29.397176   \n",
       "4  ...   23.865224   26.701340    20.760923   32.722691   28.540895   \n",
       "\n",
       "   YWHAZ~P63104      group  sample  scFv~Q65ZC9  subject  \n",
       "0     19.205044  Cirrhosis    AS65    27.788928      S65  \n",
       "1     22.141285  Cirrhosis    AS66    26.869972      S66  \n",
       "2     18.872159  Cirrhosis    AS67    28.069328      S67  \n",
       "3     22.216971  Cirrhosis    AS69    28.170209      S69  \n",
       "4     20.886069  Cirrhosis    AS70    28.612280      S70  \n",
       "\n",
       "[5 rows x 522 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "processed_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "processed_df = processed_df.drop(['sample', 'subject'], axis=1).set_index('group').apply(zscore).reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### In order to find clusters of proteins, we access the report and the protein-protein correlation network as a dictionary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['0_date', '0~proteomics_pipeline~cytoscape_network', '10~ranking_ranking_with_markers~basicTable', '11~stratification_description~description', '12~stratification_pca~pca', '13~regulation_description~description', '14~regulation_samr~basicTable', '15~regulation_samr~volcanoplot', '16~correlation_correlation~network', '17~interaction_network~network', '18~complex_associations~basicTable', '19~drug_associations~basicTable', '1~overview statistics_summary~multiTable', '20~disease_associations~basicTable', '21~literature_associations_publications_abstracts~basicTable', '22~literature_associations_publications_abstracts~wordcloud', '23~go_enrichment_Biological_processes_regulation_enrichment~basicTable', '24~pathway_enrichment_Pathways_regulation_enrichment~basicTable', '2~peptides~barplot', '3~peptides~basicTable', '4~proteins~barplot', '5~proteins~basicTable', '6~modifications~facetplot', '7~modifications~basicTable', '8~coefficient_variation_coefficient_of_variation~scatterplot_matrix', '9~ranking_ranking_with_markers~ranking'])"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proteomics_report = my_project.get_dataset('proteomics').report\n",
    "proteomics_report.list_plots()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "correlation_net_dict = proteomics_report.get_plot('16~correlation_correlation~network')[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### To convert the dictionary into a network, we access the json version within the dictionary and convert it using the networkX package."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "correlation_net = json_graph.node_link_graph(correlation_net_dict['net_json'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Now that we have a network with proteins colored by cluster, we can convert this information into a dataframe to be used in this Jupyter Notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "correlation_df = pd.DataFrame.from_dict(correlation_net.nodes(data=True))\n",
    "correlation_df = correlation_df[0].to_frame().join(correlation_df[1].apply(pd.Series))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "correlation_df.columns = ['identifier', 'degree', 'radius', 'color', 'cluster']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Since the correlation network was generated using cut-off , not all the proteins in the processed dataframe are part of a cluster, therefore we filter the processed dataframe and keep only the proteins that are present in the correlation clusters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "min_val = processed_df._get_numeric_data().min().min().round()\n",
    "max_val = processed_df._get_numeric_data().max().max().round()\n",
    "processed_df = processed_df[list(correlation_df.identifier) + ['group']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Ready! To build the parallel plot, we create a dictionary with the clusters and respectives colors, and filter the processed dataframe to include only the proteins in a specific cluster.\n",
    "Using the Jupyter Widgets **interact** function, we can make the plot interactive and allow the visualization of a cluster selected by the user."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.core.display import display, HTML"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6cd2ff7803ea4aec8fb7f6974f241e09",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "interactive(children=(Dropdown(description='cluster', options=(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 1…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "@interact\n",
    "def plot_parallel_plot(cluster=correlation_df.cluster.unique()):\n",
    "    cluster_colors = dict(zip(correlation_df.cluster, correlation_df.color))\n",
    "    clusters = correlation_df.groupby('cluster')\n",
    "    identifiers = clusters.get_group(cluster)['identifier'].tolist()\n",
    "    title= \"Parallel plot cluster: {}\".format(cluster)\n",
    "    df = processed_df.set_index('group')[identifiers].reset_index()\n",
    "    figure = plots.get_parallel_plot(df, identifier=cluster, args={'color':cluster_colors[cluster],'group':'group', \n",
    "                                                                          'title':title,\n",
    "                                                                          'zscore':False})\n",
    "    display(HTML(\"<p>{}</p>\".format(\",\".join(identifiers))))\n",
    "    iplot(figure.figure)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
